File: cardre/store/schema.py
"""SQL schema for the Cardre v2 project store.

Hard break from v1: ``V2_STORE_SCHEMA_VERSION = 100``.
No MIGRATIONS_SQL, no backfill.  ``store_meta`` is a hard-version-check,
not a migration table.
"""

# Explicit v2 identity for the new store in ``cardre.store.db``.
V2_STORE_SCHEMA_FAMILY = "cardre-v2"
V2_STORE_SCHEMA_VERSION = 100

# ---------------------------------------------------------------------------
# Core tables
# ---------------------------------------------------------------------------

SCHEMA_SQL = """
CREATE TABLE IF NOT EXISTS store_meta (
    key TEXT PRIMARY KEY,
    value TEXT NOT NULL
);

CREATE TABLE IF NOT EXISTS projects (
    project_id TEXT PRIMARY KEY,
    name TEXT NOT NULL,
    created_at TEXT NOT NULL,
    cardre_version TEXT NOT NULL,
    metadata_json TEXT NOT NULL DEFAULT '{}'
);

CREATE TABLE IF NOT EXISTS plans (
    plan_id TEXT PRIMARY KEY,
    project_id TEXT NOT NULL REFERENCES projects(project_id) ON DELETE CASCADE,
    name TEXT NOT NULL,
    created_at TEXT NOT NULL,
    metadata_json TEXT NOT NULL DEFAULT '{}'
);

CREATE TABLE IF NOT EXISTS plan_versions (
    plan_version_id TEXT PRIMARY KEY,
    plan_id TEXT NOT NULL REFERENCES plans(plan_id) ON DELETE CASCADE,
    version_number INTEGER NOT NULL,
    is_committed INTEGER NOT NULL DEFAULT 0,
    created_at TEXT NOT NULL,
    description TEXT NOT NULL DEFAULT '',
    metadata_json TEXT NOT NULL DEFAULT '{}',
    UNIQUE(plan_id, version_number)
);

CREATE TABLE IF NOT EXISTS plan_steps (
    step_id TEXT NOT NULL,
    plan_version_id TEXT NOT NULL REFERENCES plan_versions(plan_version_id) ON DELETE CASCADE,
    node_type TEXT NOT NULL,
    node_version TEXT NOT NULL,
    category TEXT NOT NULL,
    params_json TEXT NOT NULL,
    params_hash TEXT NOT NULL,
    branch_label TEXT NOT NULL DEFAULT '',
    position INTEGER NOT NULL,
    canonical_step_id TEXT NOT NULL DEFAULT '',
    branch_id TEXT,
    PRIMARY KEY (plan_version_id, step_id)
);

-- Replaces plan_steps.parent_step_ids_json (v1)
CREATE TABLE IF NOT EXISTS plan_step_edges (
    plan_version_id TEXT NOT NULL,
    parent_step_id TEXT NOT NULL,
    child_step_id TEXT NOT NULL,
    edge_order INTEGER NOT NULL DEFAULT 0,
    PRIMARY KEY (plan_version_id, parent_step_id, child_step_id),
    FOREIGN KEY(plan_version_id, parent_step_id)
        REFERENCES plan_steps(plan_version_id, step_id) ON DELETE CASCADE,
    FOREIGN KEY(plan_version_id, child_step_id)
        REFERENCES plan_steps(plan_version_id, step_id) ON DELETE CASCADE
);

CREATE TABLE IF NOT EXISTS runs (
    run_id TEXT PRIMARY KEY,
    plan_version_id TEXT NOT NULL REFERENCES plan_versions(plan_version_id) ON DELETE CASCADE,
    status TEXT NOT NULL,
    run_scope TEXT NOT NULL DEFAULT 'full_plan',
    branch_id TEXT,
    target_step_id TEXT,
    force INTEGER NOT NULL DEFAULT 0,
    requested_by TEXT,
    request_id TEXT,
    created_at TEXT NOT NULL,
    queued_at TEXT,
    started_at TEXT NOT NULL,
    finished_at TEXT,
    heartbeat_at TEXT,
    metadata_json TEXT NOT NULL DEFAULT '{}'
);

CREATE TABLE IF NOT EXISTS run_steps (
    run_step_id TEXT PRIMARY KEY,
    run_id TEXT NOT NULL REFERENCES runs(run_id) ON DELETE CASCADE,
    step_id TEXT NOT NULL,
    plan_version_id TEXT NOT NULL,
    status TEXT NOT NULL,
    started_at TEXT NOT NULL,
    finished_at TEXT,
    execution_fingerprint_json TEXT NOT NULL,
    warnings_json TEXT NOT NULL DEFAULT '[]',
    errors_json TEXT NOT NULL DEFAULT '[]'
);

CREATE TABLE IF NOT EXISTS artifacts (
    artifact_id TEXT PRIMARY KEY,
    artifact_type TEXT NOT NULL,
    role TEXT NOT NULL,
    path TEXT NOT NULL,
    physical_hash TEXT NOT NULL,
    logical_hash TEXT NOT NULL,
    media_type TEXT NOT NULL,
    created_at TEXT NOT NULL,
    metadata_json TEXT NOT NULL DEFAULT '{}'
);

CREATE TABLE IF NOT EXISTS artifact_lineage (
    lineage_id TEXT PRIMARY KEY,
    run_id TEXT NOT NULL REFERENCES runs(run_id) ON DELETE CASCADE,
    run_step_id TEXT NOT NULL REFERENCES run_steps(run_step_id) ON DELETE CASCADE,
    plan_version_id TEXT NOT NULL REFERENCES plan_versions(plan_version_id) ON DELETE CASCADE,
    step_id TEXT NOT NULL,
    branch_id TEXT,
    artifact_id TEXT NOT NULL REFERENCES artifacts(artifact_id) ON DELETE CASCADE,
    direction TEXT NOT NULL,
    created_at TEXT NOT NULL,
    UNIQUE(run_step_id, artifact_id, direction)
);
"""

# ---------------------------------------------------------------------------
# Evidence tables — two-level model
# ---------------------------------------------------------------------------

EVIDENCE_TABLES_SQL = """
CREATE TABLE IF NOT EXISTS evidence_edges (
    evidence_edge_id TEXT PRIMARY KEY,
    run_id TEXT NOT NULL REFERENCES runs(run_id) ON DELETE CASCADE,
    run_step_id TEXT NOT NULL REFERENCES run_steps(run_step_id) ON DELETE CASCADE,
    plan_version_id TEXT NOT NULL,
    step_id TEXT NOT NULL,
    parent_step_id TEXT NOT NULL,
    source_run_id TEXT NOT NULL,
    source_run_step_id TEXT NOT NULL REFERENCES run_steps(run_step_id) ON DELETE CASCADE,
    policy TEXT NOT NULL,
    source_label TEXT NOT NULL,
    is_reused INTEGER NOT NULL,
    is_stale INTEGER NOT NULL,
    stale_reason TEXT,
    created_at TEXT NOT NULL,
    UNIQUE(run_step_id, parent_step_id, source_run_step_id)
);

CREATE TABLE IF NOT EXISTS evidence_artifacts (
    evidence_artifact_id TEXT PRIMARY KEY,
    evidence_edge_id TEXT NOT NULL REFERENCES evidence_edges(evidence_edge_id) ON DELETE CASCADE,
    artifact_id TEXT NOT NULL REFERENCES artifacts(artifact_id) ON DELETE CASCADE,
    role TEXT NOT NULL,
    created_at TEXT NOT NULL,
    UNIQUE(evidence_edge_id, artifact_id)
);
"""

# ---------------------------------------------------------------------------
# Branch / governance tables
# ---------------------------------------------------------------------------

BRANCH_TABLES_SQL = """
CREATE TABLE IF NOT EXISTS plan_branches (
    branch_id TEXT PRIMARY KEY,
    project_id TEXT NOT NULL,
    plan_id TEXT NOT NULL,
    name TEXT NOT NULL,
    description TEXT,
    branch_type TEXT NOT NULL,
    status TEXT NOT NULL DEFAULT 'active',
    base_branch_id TEXT REFERENCES plan_branches(branch_id),
    base_plan_version_id TEXT NOT NULL REFERENCES plan_versions(plan_version_id),
    head_plan_version_id TEXT NOT NULL REFERENCES plan_versions(plan_version_id),
    branch_point_step_id TEXT,
    branch_point_canonical_step_id TEXT,
    segment_filter_spec_json TEXT,
    created_reason TEXT NOT NULL,
    created_at TEXT NOT NULL,
    updated_at TEXT NOT NULL,
    archived_at TEXT,
    FOREIGN KEY(project_id) REFERENCES projects(project_id),
    FOREIGN KEY(plan_id) REFERENCES plans(plan_id)
);

CREATE TABLE IF NOT EXISTS branch_step_map (
    branch_step_map_id TEXT PRIMARY KEY,
    branch_id TEXT NOT NULL,
    plan_version_id TEXT NOT NULL REFERENCES plan_versions(plan_version_id),
    canonical_step_id TEXT NOT NULL,
    step_id TEXT NOT NULL,
    source_branch_id TEXT REFERENCES plan_branches(branch_id),
    source_step_id TEXT,
    is_shared_upstream INTEGER NOT NULL DEFAULT 0,
    is_branch_owned INTEGER NOT NULL DEFAULT 1,
    created_at TEXT NOT NULL,
    FOREIGN KEY(branch_id) REFERENCES plan_branches(branch_id)
);

CREATE TABLE IF NOT EXISTS branch_comparisons (
    comparison_id TEXT PRIMARY KEY,
    project_id TEXT NOT NULL REFERENCES projects(project_id),
    plan_id TEXT NOT NULL REFERENCES plans(plan_id),
    baseline_branch_id TEXT NOT NULL,
    comparison_spec_json TEXT NOT NULL,
    latest_snapshot_id TEXT,
    latest_ready INTEGER,
    latest_readiness_json TEXT,
    created_at TEXT NOT NULL,
    created_reason TEXT,
    FOREIGN KEY(baseline_branch_id) REFERENCES plan_branches(branch_id)
);

CREATE TABLE IF NOT EXISTS comparison_challenger_branches (
    comparison_id TEXT NOT NULL REFERENCES branch_comparisons(comparison_id) ON DELETE CASCADE,
    branch_id TEXT NOT NULL REFERENCES plan_branches(branch_id) ON DELETE CASCADE,
    position INTEGER NOT NULL DEFAULT 0,
    PRIMARY KEY (comparison_id, branch_id)
);

CREATE TABLE IF NOT EXISTS branch_comparison_snapshots (
    comparison_snapshot_id TEXT PRIMARY KEY,
    comparison_id TEXT NOT NULL,
    project_id TEXT NOT NULL,
    plan_id TEXT NOT NULL,
    comparison_artifact_id TEXT NOT NULL REFERENCES artifacts(artifact_id),
    readiness_json TEXT NOT NULL,
    created_at TEXT NOT NULL,
    created_reason TEXT,
    FOREIGN KEY(comparison_id) REFERENCES branch_comparisons(comparison_id)
);

CREATE TABLE IF NOT EXISTS comparison_snapshot_plan_versions (
    comparison_snapshot_id TEXT NOT NULL REFERENCES branch_comparison_snapshots(comparison_snapshot_id) ON DELETE CASCADE,
    plan_version_id TEXT NOT NULL REFERENCES plan_versions(plan_version_id),
    branch_id TEXT REFERENCES plan_branches(branch_id),
    PRIMARY KEY (comparison_snapshot_id, plan_version_id)
);
"""

# ---------------------------------------------------------------------------
# Annotation / champion / export tables
# ---------------------------------------------------------------------------

ANNOTATION_TABLES_SQL = """
CREATE TABLE IF NOT EXISTS step_annotations (
    annotation_id TEXT PRIMARY KEY,
    step_id TEXT NOT NULL,
    plan_version_id TEXT NOT NULL,
    kind TEXT NOT NULL,
    actor TEXT NOT NULL,
    payload_json TEXT NOT NULL DEFAULT '{}',
    created_at TEXT NOT NULL,
    FOREIGN KEY(plan_version_id, step_id) REFERENCES plan_steps(plan_version_id, step_id) ON DELETE CASCADE
);

CREATE TABLE IF NOT EXISTS champion_assignments (
    champion_assignment_id TEXT PRIMARY KEY,
    project_id TEXT NOT NULL REFERENCES projects(project_id),
    plan_id TEXT NOT NULL REFERENCES plans(plan_id),
    scope_type TEXT NOT NULL,
    scope_key TEXT NOT NULL,
    champion_branch_id TEXT NOT NULL,
    comparison_id TEXT NOT NULL,
    comparison_snapshot_id TEXT NOT NULL,
    comparison_artifact_id TEXT NOT NULL REFERENCES artifacts(artifact_id),
    selected_plan_version_id TEXT NOT NULL REFERENCES plan_versions(plan_version_id),
    assigned_reason TEXT NOT NULL,
    assigned_by TEXT,
    assigned_at TEXT NOT NULL,
    superseded_at TEXT,
    superseded_by_assignment_id TEXT REFERENCES champion_assignments(champion_assignment_id),
    FOREIGN KEY(champion_branch_id) REFERENCES plan_branches(branch_id),
    FOREIGN KEY(comparison_id) REFERENCES branch_comparisons(comparison_id),
    FOREIGN KEY(comparison_snapshot_id) REFERENCES branch_comparison_snapshots(comparison_snapshot_id)
);
"""

# ---------------------------------------------------------------------------
# Manual-binning review + diagnostics + store_meta
# ---------------------------------------------------------------------------

REVIEW_TABLES_SQL = """
CREATE TABLE IF NOT EXISTS diagnostics (
    diagnostic_id TEXT PRIMARY KEY,
    run_id TEXT NOT NULL REFERENCES runs(run_id) ON DELETE CASCADE,
    code TEXT NOT NULL,
    message TEXT NOT NULL,
    source TEXT,
    severity TEXT NOT NULL DEFAULT 'error',
    context_json TEXT NOT NULL DEFAULT '{}',
    created_at TEXT NOT NULL
);

CREATE TABLE IF NOT EXISTS manual_binning_reviews (
    review_id TEXT PRIMARY KEY,
    plan_version_id TEXT NOT NULL,
    step_id TEXT NOT NULL,
    status TEXT NOT NULL,
    reviewer_notes TEXT NOT NULL DEFAULT '',
    affected_downstream_step_ids_json TEXT NOT NULL DEFAULT '[]',
    created_at TEXT NOT NULL,
    updated_at TEXT NOT NULL,
    FOREIGN KEY(plan_version_id, step_id) REFERENCES plan_steps(plan_version_id, step_id) ON DELETE CASCADE
);
"""

# ---------------------------------------------------------------------------
# Exports table
# ---------------------------------------------------------------------------

EXPORTS_TABLE_SQL = """
CREATE TABLE IF NOT EXISTS exports (
    export_id TEXT PRIMARY KEY,
    run_id TEXT NOT NULL REFERENCES runs(run_id) ON DELETE CASCADE,
    export_type TEXT NOT NULL,
    path TEXT NOT NULL,
    size_bytes INTEGER NOT NULL DEFAULT 0,
    created_at TEXT NOT NULL
);
"""

# ---------------------------------------------------------------------------
# Indexes
# ---------------------------------------------------------------------------

INDEXES_SQL = """
CREATE INDEX IF NOT EXISTS idx_runs_plan_version_status
    ON runs(plan_version_id, status, started_at DESC);
CREATE INDEX IF NOT EXISTS idx_run_steps_run_id
    ON run_steps(run_id);
CREATE INDEX IF NOT EXISTS idx_run_steps_pv_step_status
    ON run_steps(plan_version_id, step_id, status, started_at DESC);
CREATE INDEX IF NOT EXISTS idx_plans_project_created
    ON plans(project_id, created_at);
CREATE INDEX IF NOT EXISTS idx_plan_versions_plan_version
    ON plan_versions(plan_id, version_number);
CREATE INDEX IF NOT EXISTS idx_plan_step_edges_child
    ON plan_step_edges(plan_version_id, child_step_id);
CREATE INDEX IF NOT EXISTS idx_plan_step_edges_parent
    ON plan_step_edges(plan_version_id, parent_step_id);
CREATE INDEX IF NOT EXISTS idx_plan_branches_project_plan_status
    ON plan_branches(project_id, plan_id, status, created_at);
CREATE INDEX IF NOT EXISTS idx_branch_step_map_branch_pv
    ON branch_step_map(branch_id, plan_version_id);
CREATE INDEX IF NOT EXISTS idx_artifacts_type_role
    ON artifacts(artifact_type, role);
CREATE INDEX IF NOT EXISTS idx_lineage_artifact ON artifact_lineage(artifact_id);
CREATE INDEX IF NOT EXISTS idx_lineage_run_direction ON artifact_lineage(run_id, direction);
CREATE INDEX IF NOT EXISTS idx_lineage_step_direction ON artifact_lineage(step_id, direction);
CREATE INDEX IF NOT EXISTS idx_lineage_pv_step ON artifact_lineage(plan_version_id, step_id);
CREATE INDEX IF NOT EXISTS idx_lineage_run_step ON artifact_lineage(run_step_id, direction);
CREATE INDEX IF NOT EXISTS idx_lineage_branch_direction ON artifact_lineage(branch_id, direction);
CREATE INDEX IF NOT EXISTS idx_evidence_edges_run_step
    ON evidence_edges(run_step_id);
CREATE INDEX IF NOT EXISTS idx_evidence_edges_pv_step
    ON evidence_edges(plan_version_id, step_id);
CREATE INDEX IF NOT EXISTS idx_evidence_edges_parent
    ON evidence_edges(plan_version_id, parent_step_id);
CREATE INDEX IF NOT EXISTS idx_evidence_edges_source_step
    ON evidence_edges(source_run_step_id);
CREATE INDEX IF NOT EXISTS idx_evidence_artifacts_artifact
    ON evidence_artifacts(artifact_id);
CREATE INDEX IF NOT EXISTS idx_evidence_artifacts_edge_role
    ON evidence_artifacts(evidence_edge_id, role);
CREATE INDEX IF NOT EXISTS idx_diagnostics_run
    ON diagnostics(run_id);
CREATE INDEX IF NOT EXISTS idx_comparison_challenger_branches_comparison
    ON comparison_challenger_branches(comparison_id);
CREATE INDEX IF NOT EXISTS idx_comparison_snapshot_pv_snapshot
    ON comparison_snapshot_plan_versions(comparison_snapshot_id);
CREATE INDEX IF NOT EXISTS idx_step_annotations_pv_step
    ON step_annotations(plan_version_id, step_id);
CREATE INDEX IF NOT EXISTS idx_manual_binning_reviews_pv_step
    ON manual_binning_reviews(plan_version_id, step_id);
CREATE INDEX IF NOT EXISTS idx_champion_assignments_plan_superseded
    ON champion_assignments(plan_id, superseded_at);
CREATE INDEX IF NOT EXISTS idx_exports_run
    ON exports(run_id);
"""

# ---------------------------------------------------------------------------
# Full schema assembly (for fresh store creation)
# ---------------------------------------------------------------------------

ALL_TABLES_SQL = (
    SCHEMA_SQL
    + EVIDENCE_TABLES_SQL
    + BRANCH_TABLES_SQL
    + ANNOTATION_TABLES_SQL
    + REVIEW_TABLES_SQL
    + EXPORTS_TABLE_SQL
    + INDEXES_SQL
)

__all__ = [
    "ALL_TABLES_SQL",
    "ANNOTATION_TABLES_SQL",
    "BRANCH_TABLES_SQL",
    "EVIDENCE_TABLES_SQL",
    "EXPORTS_TABLE_SQL",
    "INDEXES_SQL",
    "REVIEW_TABLES_SQL",
    "SCHEMA_SQL",
    "V2_STORE_SCHEMA_FAMILY",
    "V2_STORE_SCHEMA_VERSION",
]

File: frontend/src/api/schema.ts
/**
 * TypeScript types matching the Cardre v2 full API surface.
 * Phase 4 — generated manually to match Pydantic models in cardre/api/schemas.py.
 *
 * No frontend/src/types.ts — use generated components types only.
 */

// ---------------------------------------------------------------------------
// Error envelope
// ---------------------------------------------------------------------------

export interface ErrorDetail {
  code: string;
  message: string;
  context: Record<string, unknown>;
}

export interface ErrorResponse {
  detail: ErrorDetail;
}

// ---------------------------------------------------------------------------
// Health
// ---------------------------------------------------------------------------

export interface HealthResponse {
  status: string;
  version: string;
}

// ---------------------------------------------------------------------------
// Projects
// ---------------------------------------------------------------------------

export interface ProjectResponse {
  project_id: string;
  name: string;
  created_at: string;
  cardre_version: string;
}

export interface ProjectListResponse {
  projects: ProjectResponse[];
}

export interface ProjectCreateRequest {
  name: string;
}

// ---------------------------------------------------------------------------
// Plans
// ---------------------------------------------------------------------------

export interface PlanResponse {
  plan_id: string;
  project_id: string;
  name: string;
  created_at: string;
}

export interface PlanListResponse {
  plans: PlanResponse[];
}

export interface PlanCreateRequest {
  name: string;
}

export interface PlanVersionResponse {
  plan_version_id: string;
  plan_id: string;
  version_number: number;
  is_committed: boolean;
  created_at: string;
  description: string;
}

export interface PlanVersionListResponse {
  versions: PlanVersionResponse[];
}

export interface PlanVersionUpdate {
  description?: string;
}

export interface PlanStepResponse {
  step_id: string;
  plan_version_id: string;
  node_type: string;
  node_version: string;
  category: string;
  params: Record<string, unknown>;
  params_hash: string;
  parent_step_ids: string[];
  branch_label: string;
  position: number;
  canonical_step_id: string;
  branch_id: string | null;
}

// ---------------------------------------------------------------------------
// Runs
// ---------------------------------------------------------------------------

export interface RunResponse {
  run_id: string;
  plan_version_id: string;
  status: string;
  started_at: string;
  finished_at: string | null;
  step_count: number;
  branch_id: string | null;
  executed_step_ids: string[];
  diagnostics: Record<string, unknown>[];
  latest_error: Record<string, unknown> | null;
  heartbeat_at: string | null;
  is_stale: boolean;
}

export interface RunListResponse {
  runs: RunResponse[];
}

export interface RunCreateRequest {
  plan_version_id: string;
  force?: boolean;
  sync?: boolean;
}

export interface RunStepResponse {
  run_step_id: string;
  run_id: string;
  step_id: string;
  plan_version_id: string;
  status: string;
  started_at: string;
  finished_at: string | null;
  execution_fingerprint: Record<string, unknown>;
  warnings: Record<string, unknown>[];
  errors: Record<string, unknown>[];
}

// ---------------------------------------------------------------------------
// Evidence
// ---------------------------------------------------------------------------

export interface EvidenceEdgeResponse {
  evidence_edge_id: string;
  run_id: string;
  run_step_id: string;
  plan_version_id: string;
  step_id: string;
  parent_step_id: string;
  source_run_id: string;
  source_run_step_id: string;
  policy: string;
  source_label: string;
  is_reused: boolean;
  is_stale: boolean;
  stale_reason: string | null;
  created_at: string;
}

export interface EvidenceArtifactResponse {
  evidence_artifact_id: string;
  evidence_edge_id: string;
  artifact_id: string;
  role: string;
  created_at: string;
}

export interface ResolvedEvidenceResponse {
  run_step_id: string;
  edges: EvidenceEdgeResponse[];
  artifacts: EvidenceArtifactResponse[];
}

export interface StalenessExplanationResponse {
  step_id: string;
  status: "fresh" | "stale" | "missing";
  upstream_changes: Record<string, boolean>;
  missing_evidence: string[];
}

// ---------------------------------------------------------------------------
// Artifacts
// ---------------------------------------------------------------------------

export interface ArtifactResponse {
  artifact_id: string;
  artifact_type: string;
  role: string;
  path: string;
  physical_hash: string;
  logical_hash: string;
  media_type: string;
  created_at: string;
}

export interface ArtifactListResponse {
  artifacts: ArtifactResponse[];
}

// ---------------------------------------------------------------------------
// Manual Binning
// ---------------------------------------------------------------------------

export interface ManualBinningReviewResponse {
  review_id: string;
  plan_version_id: string;
  step_id: string;
  status: string;
  reviewer_notes: string;
  affected_downstream_step_ids: string[];
  created_at: string;
  updated_at: string;
}

export interface ManualBinningReviewUpdate {
  status?: string;
  reviewer_notes?: string;
}

export interface ManualBinningEditRequest {
  plan_version_id: string;
  step_id: string;
  overrides: Record<string, unknown>[];
  reviewer_notes: string;
  status: string;
  affected_downstream_step_ids: string[];
}

export interface ManualBinningEditResponse {
  new_plan_version_id: string;
  review_id: string;
  affected_step_ids: string[];
}

export interface ManualBinningPreviewRequest {
  variable_data: Record<string, unknown>;
}

export interface ManualBinningPreviewResponse {
  woe_by_bin: Record<string, unknown>[];
  iv: number;
  event_rate_by_bin: Record<string, unknown>[];
}

// ---------------------------------------------------------------------------
// Branches (governance-gated)
// ---------------------------------------------------------------------------

export interface BranchResponse {
  branch_id: string;
  project_id: string;
  plan_id: string;
  name: string;
  description: string | null;
  branch_type: string;
  status: string;
  base_branch_id: string | null;
  base_plan_version_id: string;
  head_plan_version_id: string;
  branch_point_step_id: string | null;
  branch_point_canonical_step_id: string | null;
  created_reason: string;
  created_at: string;
  updated_at: string;
}

export interface BranchListResponse {
  branches: BranchResponse[];
}

export interface BranchCreateRequest {
  plan_id: string;
  name: string;
  branch_type: string;
  base_plan_version_id: string;
  head_plan_version_id: string;
  description?: string | null;
  base_branch_id?: string | null;
  branch_point_step_id?: string | null;
  created_reason?: string;
}

// ---------------------------------------------------------------------------
// Comparisons (governance-gated)
// ---------------------------------------------------------------------------

export interface ComparisonResponse {
  comparison_id: string;
  project_id: string;
  plan_id: string;
  baseline_branch_id: string;
  created_at: string;
  latest_ready: boolean | null;
}

export interface ComparisonListResponse {
  comparisons: ComparisonResponse[];
}

// ---------------------------------------------------------------------------
// Champion (governance-gated)
// ---------------------------------------------------------------------------

export interface ChampionAssignmentResponse {
  champion_assignment_id: string;
  project_id: string;
  plan_id: string;
  champion_branch_id: string;
  selected_plan_version_id: string;
  assigned_at: string;
  superseded_at: string | null;
}

export interface ChampionResponse {
  assignment: ChampionAssignmentResponse | null;
}

// ---------------------------------------------------------------------------
// Node types
// ---------------------------------------------------------------------------

export interface NodeTypeResponse {
  node_type: string;
  display_name: string;
  description: string;
  category: string;
  tier: string;
  has_params: boolean;
}

export interface NodeTypeListResponse {
  node_types: NodeTypeResponse[];
}

// ---------------------------------------------------------------------------
// Exports
// ---------------------------------------------------------------------------

export interface ExportResponse {
  export_id: string;
  run_id: string;
  export_type: string;
  path: string;
  created_at: string;
  size_bytes: number;
}

export interface ExportListResponse {
  exports: ExportResponse[];
}

// ---------------------------------------------------------------------------
// Reports
// ---------------------------------------------------------------------------

export interface ReportResponse {
  report_id: string;
  run_id: string | null;
  report_type: string;
  path: string;
  created_at: string;
}

export interface ReportListResponse {
  reports: ReportResponse[];
}

// ---------------------------------------------------------------------------
// Canonical error codes
// ---------------------------------------------------------------------------

export const ErrorCodes = {
  SIDECAR_UNREACHABLE: "SIDECAR_UNREACHABLE",
  REQUEST_TIMEOUT: "REQUEST_TIMEOUT",
  REQUEST_ABORTED: "REQUEST_ABORTED",
  EMPTY_OK_BODY: "EMPTY_OK_BODY",
  EMPTY_ERROR_RESPONSE: "EMPTY_ERROR_RESPONSE",
  MALFORMED_JSON_RESPONSE: "MALFORMED_JSON_RESPONSE",
  HTML_ERROR_RESPONSE: "HTML_ERROR_RESPONSE",
  NON_JSON_ERROR_RESPONSE: "NON_JSON_ERROR_RESPONSE",
  GOVERNANCE_DISABLED: "GOVERNANCE_DISABLED",
  PROJECT_NOT_FOUND: "PROJECT_NOT_FOUND",
  PLAN_NOT_FOUND: "PLAN_NOT_FOUND",
  PLAN_VERSION_NOT_FOUND: "PLAN_VERSION_NOT_FOUND",
  PLAN_VERSION_IMMUTABLE: "PLAN_VERSION_IMMUTABLE",
  RUN_NOT_FOUND: "RUN_NOT_FOUND",
  STEP_NOT_FOUND: "STEP_NOT_FOUND",
  ARTIFACT_NOT_FOUND: "ARTIFACT_NOT_FOUND",
  BRANCH_NOT_FOUND: "BRANCH_NOT_FOUND",
} as const;
